### ISSUE TITLE
Can't .DeleteRealm() Even After .Dispose()

### ISSUE THREAD

**Issue by mathaou** [realm/realm-dotnet#1970]
## Goals

Trying to replace a .realm file used by a configuration offline. 

## Expected Results

After I .Dispose() the instance, I expect .DeleteRealm() to remove all files associated with it.

## Actual Results

Error thrown. File in use by another process.

## Steps to Reproduce

Create a Realm instance, attach a .RealmChanged handler, and get a reference to the .All<T>.

## Code Sample

```
Uri.TryCreate(realmPath, UriKind.Relative, out var outPathUri);
RealmConfigurationBase configuration = new FullSyncConfiguration(outPathUri, _currentUser, _realmFile)
{
    ObjectClasses = typesInSelectedRealm,
    SchemaVersion = 1
};

_realmInstance = Realm.GetInstance(configuration);

if(_realmInstance != null) _realmInstance.RealmChanged += LoadDataOnChange;
```
Trying to copy .realm file with same format from a USB Drive
```
if(_realmInstance != null) 
{
    _realmInstance.RealmChanged -= LoadDataOnChange; // figure it's good to clear all external references to Realm object
    _realmInstance.Dispose();
    GC.Collect();
    Thread.Sleep(1000); // both of these are me trying to just make darn sure that everything is cleaned up and has had enough time to do so
    Realm.DeleteRealm(_realmInstance.Config); // error
}
```
Error thrown: The process cannot access the file '...\test.realm' because it is being used by another process.

I've already seen this thread, which is the closest I've found to my issue and a possible solution, although it was for IOS so I'm not sure how much is applicable to my situation.

## Version of Realm and Tooling

- Realm Object Server Version: ? (you can see this as well in Studio once connected to a Realm instance) 4.2.0
- Flavor:
  - [X] Realm Cloud
  - [ ] Self-Hosted
- Server OS & Version: ? (e.g. CentOS 6) Windows 10 Enterprise 1806
- Client SDK Version: ? 4.2.0
- Client OS & Version: ? Windows 10 Enterprise 1806

**Comment by mathaou** [comment]
[SampleUWP.zip](https://github.com/realm/realm-dotnet/files/4344758/SampleUWP.zip)

**Comment by fealebenpae** [comment]
Hey @mathaou,

When you open a synchronized realm the framework spawns a worker thread which runs the sync session. By default after you close the last instance of a synchronized realm file its sync session lingers a bit more in the worker thread to make sure that changes are propagated.

It seems like you're opening a realm and then shortly thereafter disposing it and trying to delete it. What I think is happening is that the sync session is still performing synchronization in the background, because when you open a synchronized realm with `Realm.GetInstance()` you get a Realm instance right away but it doesn't hold any data until the sync session is completely done synchronizing.

My suggestion is to try using `Realm.GetInstanceAsync()` instead - the difference is that this method will wait for initial synchronization to be completed before returning the Realm instance, which means that after you dispose the realm instance the sync session should release the realm file right away as it doesn't actually have anything to keep synchronizing (`Realm.GetInstanceAsync()` already took care of the initial synchronization).

Alternatively, you can call `realm.GetSession().Stop()` before disposing your realm instance - this will stop the sync session which means it won't linger after you close the realm to finish synchronization and the file should be able to be deleted.

Out of curiosity, under what circumstances do you need to delete a synchronized realm file? If it's a file you'll be opening again that'll just mean it'll need to be downloaded from the server all over again as opposed to just resuming from where it left off.

**Comment by mathaou** [comment]
@fealebenpae You almost solved my issue! Not sure why, but I can successfully overwrite the file but only once...

`await Realm.GetInstanceAsync(configuration)` threw an Exception for me. Just said to look for `InnerException` for more info, but I can't find any info on that. So close! This was the secret sauce:

```
_realmInstance.GetSession().Stop();
_realmInstance.Dispose();
```

Now I just need to find why I can only do overwrite once...

**Comment by mathaou** [comment]
[SampleUWP.zip](https://github.com/realm/realm-dotnet/files/4344872/SampleUWP.zip)

**Comment by mathaou** [comment]
It works once and then I just restart the app. Good enough for my purposes.

**Comment by mathaou** [comment]
Reopening the issue.

The application I'm working on needs to be able to handle devices in an offline environment, but we still want to use Realm. I have changed the FullSyncConfiguration to a RealmConfiguration, so there is no longer a Session to stop. I'm also successfully getting the instance asynchronously, but the issue remains with not being able to delete the Realm even after I have disposed the instance.
